edfio/edf_signal.py

```diff
@@ -340,12 +333,79 @@ def data(self) -> npt.NDArray[np.float64]:
                     f"Physical minimum equals physical maximum ({self.physical_min}) for {self.label}, returning uncalibrated signal."
                 )
         except ValueError:
-            data = self.digital.astype(np.float64)
+            data = digital.astype(np.float64)
         else:
-            data = (self.digital + offset) * gain
+            data = (digital + offset) * gain
         data.setflags(write=False)
         return data
 
+    @property
+    def data(self) -> npt.NDArray[np.float64]:
+        """
+        Numpy array containing the physical signal values as floats.
+
+        To simplify avoiding inconsistencies between signal data and header fields,
+        individual values in the returned array can not be modified. Use
+        :meth:`EdfSignal.update_data` to overwrite with new physical data.
+        """
+        return self._calibrate(self.digital)
+
+    def get_digital_slice(
+        self, start_second: float, duration: float
+    ) -> npt.NDArray[np.int16]:
+        """
+        Get a slice of the digital signal values.
+
+        If the signal has not been loaded into memory so far, only the requested slice will be read.
+
+        Parameters
+        ----------
+        start_second : float
+            The start of the slice in seconds.
+        duration : float
+            The duration of the slice in seconds.
+        """
+        if duration < 0:
+            raise ValueError("Invalid slice: Duration must be non-negative")
+        if start_second < 0:
+            raise ValueError("Invalid slice: Start second must be non-negative")
+        start_index = round(start_second * self.sampling_frequency)
+        end_index = round((start_second + duration) * self.sampling_frequency)
+        if self._digital is not None:
+            return self._digital[start_index:end_index]
+        if self._lazy_loader is None:
+            raise ValueError("Signal data not set")
+        first_data_record = start_index // self.samples_per_data_record
+        last_data_record = (end_index - 1) // self.samples_per_data_record + 1
+        try:
+            digital_portion = self._lazy_loader.load(
+                first_data_record, last_data_record
+            )
+        except Exception as e:
```

Comment:
At this point, there are two likely reasons for an exception being thrown in the `load` method:
- Some kind of IO error
- Trying to access data records beyond the end of the file
How about throwing a very specific exception in `load` when the second case happens, catching it here, and throwing a ValueError with a with a message like "Invalid slice: End second must be less than or equal the duration of the file"?
This would sort of address the previous comment as well.

Reply:
Yes, that would be simpler and would produce the exact same result. It's just my own compulsivity that's prevented me from doing that. I will try to explain:
- I know that at the moment there is only one place where the LazyLoader class is used to read a segment of the file, and that is in the EdfSignal's get_slice method.
- So theoretically, I can throw an exception with a text that is appropriate for the "callee".
- I like to avoid doing that, because at some point the LazyLoader might be called the same way in a different context, and the exception text might not be appropriate for that case.

So that's why I personally prefer to raise an exception that only assumes knowledge from the local context where it is raised and then catch it at a higher-level just to wrap it in another exception with a more appropriate error message for that context. I don't know if that makes any sense? I also don't know if doing that justifies the extra lines of code. If this sounds like complete nonsense to you, I can just throw the ValueError right away in the LazyLoader.

Reply:
I understand, thanks for the thorough explanation! I'd prefer to keep it simple for now and introduce this additional layer once we actually do need different exceptions.

Reply:
Okay, done!